### ISSUE TITLE
What does exports.boilerplate do?

### ISSUE THREAD

**Issue by Sequoia** [timoxley/adventure-map#7]

**Comment by timoxley** [comment]
This is the boilerplate content generated for the user for the exercise.

e.g.

```js
module.exports = function(arg) {
  // your solution here
}
```

**Comment by Sequoia** [comment]
mmm. but... why is it included in exports? is it consumed somewhere? oh, can users get it via `workshopname boilerplate`?

**Comment by timoxley** [comment]
it automatically dumps it into the user's solution directory.

I should probably expose this so it's less magical.

**Comment by Sequoia** [comment]
ahhh OK thanks! what's "The user's solution directory"? When I was working thru these things there wasn't really a place to put files unless you made one. Is that another new feature?

**Comment by timoxley** [comment]
yeah I cloned a feature from the webgl/shaderschool workshops, where they generate a directory for each exercise up front, that directory contains the bootstrap.js and the exercise description so they don't have to keep opening and closing the workshop to review the instructions.

**Comment by timoxley** [comment]
this _is_ an assumption though and I should make it clearer how to turn this off or customise it.

**Comment by Sequoia** [comment]
cool. I think boilerplate but it would be nice to be able to disable it, perhaps by simply not exporting a `boilerplate` property or setting `exports.boilerplate` to `false`.

I plan on adding instructions in the top-level README for my lesson-set explaining the stuff about the boilerplate code & how to get started. I think this is necessary, otherwise the "we're about to populate" message is jarring & confusing. I'll think about ways to reword it. Something like "Would you like a directory created for this exercise with a boilerplate (starting point) file? Click yes & `lesson-name/` will be created in the current directory."

**Comment by timoxley** [comment]
Sounds good!

**Comment by Sequoia** [comment]
OK assign me then & I'll get to it when I have a mo.

**Comment by timoxley** [comment]
Added you as collab :tada: 

Also note: https://github.com/timoxley/adventure-map/pull/3

**Comment by Sequoia** [comment]
Planning to prompt to create boilerplate files on first run & just leave it at that, simplifies behavior muchly.

**Comment by timoxley** [comment]
@Sequoia sounds good

**Comment by Sequoia** [comment]
Wow

**Comment by timoxley** [comment]
:fireworks:

**Comment by Sequoia** [comment]
What the heck man earlier I pocket texted someone A PHOTO and now I see apparently I pocket commented on a github issue?? Presumably via email.

OS collaboration via pocket dial-- what a time to be alive!

**Comment by timoxley** [comment]
haha hopefully your pocket can start sending PRs

**Comment by Sequoia** [comment]
I started working on this a while ago & got something mostly working, but now that I revisit it I am confused. Setup was getting run _every single time_ the user ran the adventure command; I'm trying to change that but I'm getting hung up on exactly what's supposed to run here instead:

``` js
  exercises.forEach(function(exercise) {
    shop.add(exercise.name, function() {
//???
    })
  })
```

I'll try to look at it more later when I get a chance.

**Comment by timoxley** [comment]
How to solve this I guess depends on exactly the behaviour you think is best. 

Select one or more of the following (or add other options):
- Prompt to generate all boilerplate when one is missing (current behaviour)
- Prompt to generate all boilerplate at startup
- Prompt to generate boilerplate for specific exercise, if missing
- Allow user to opt out of auto-boilerplate generation forever
- Allow user to manually generate the current boilerplate via a command
- Prompt whenever exercise selected
- Prompt only on first selection of exercise
- No prompting on selection of exercise at all
- Prompt whenever workshop is run (current behaviour)
- Prompt only on first startup of workshop

**Comment by timoxley** [comment]
Note that boilerplate generation stuff is mostly only frustrating for workshop authors while they're developing the workshop. Most users will simply get the boilerplate setup once and then work with it.

Perhaps should have a way to disable it while developing or figure out a better way to serve the workshop-developer's needs.

**Comment by Sequoia** [comment]
Nah, it's annoying if you run the workshop, switch to the new directory, run it again, then boilerplate rubs again.

Tbh my next move is to just rip it out and move on, I never finished my worksheet I started about a year ago and the boilerplate stuff seems not worth the trouble...
